### ISSUE TITLE
Are variadic functions supported?

### ISSUE THREAD

**Issue by craig65535** [ebitengine/purego#116]
The source has a comment about a special case for variadic functions -

```go
// There is a special case when the last argument of fptr is a variadic interface (or []interface}
// it will be expanded into a call to the C function as if it had the arguments in that slice.
// This means that using arg ...interface{} is like a cast to the function with the arguments inside arg.
// This is not the same as C variadic.
```

Does this mean they are not supported? I made my own library in C and am unable to find a way to get my function to see any of the variadic parameters.

```C
void printSomething(int a, const char *fmt, ...)
{
    va_list ap;
    printf("start %d\n", a);
    printf("fmt %s\n", fmt);
    va_start(ap, fmt);
    vprintf(fmt, ap);
    va_end(ap);
    printf("\n");
    printf("end %d\n", a);
}
```

Calling code:
```go
	var printSomething func(int, string, []interface{})
	purego.RegisterLibFunc(&printSomething, libcraig, `printSomething`)
	printSomething(5, `%s`, []interface{}{`foo`})
```

This prints:
```
start 5
fmt %s
(null)
end 5
```

**Comment by craig65535** [comment]
I should mention I'm using macOS (darwin/arm64).

**Comment by TotallyGamerJet** [comment]
That is correct. Variadic functions are not supported and there are currently no plans to add them.

**Comment by craig65535** [comment]
Thanks for the response.

I think the documentation for RegisterFunc/RegisterLibFunc could be improved because that wasn't clear to me. But I'm not sure what to change it to, as there is probably some use for the current arrangement with `[]interface{}` that I'm not aware of.

**Comment by TotallyGamerJet** [comment]
The current implementation is used to implement the Send method in the objc package. This way the user of the method can call any objc method without having to create a func variable link it to the C function and then call it.

**Comment by lsy88** [comment]
How do I convert Go []string to C char **

**Comment by TotallyGamerJet** [comment]
Create a slice of *byte. Convert each string to null terminated *byte. Then pass the address of the slice to the function. So the signature would be &slice[0] and the type **byte

**Comment by lsy88** [comment]
Please understand that I may be a little unclear in the description, for the case of a C function parameter char **, what type should be used in Go?

**Comment by TotallyGamerJet** [comment]
**byte like I said

**Comment by lsy88** [comment]
I'm sorry that I can't solve this matter according to your suggestion
C：
````
int test(char *path, char **result, int *len)
````
Go: 
```
var test func(path string, result **byte, rlen *int) int32
...
purego.RegisterLibFunc(&test, lic, "test")
...
func Test(path string) {
  var lenRes int
  result := make([]*byte, 0)
  _ = test(path, &result[0], &lenRes)
}
...
```
Error reported as follows：
`cannot use result (variable of type []*byte) as *_Ctype_char value in argument to (_Cfunc_GoString)`

**Comment by TotallyGamerJet** [comment]
You actually got all the code right so far. You just need to turn the *byte back into a string. There's no easy way to do this atm. There's discussion in https://github.com/ebitengine/purego/issues/121 about making it more accessible. You can also just copy the code that's inside purego/internal/strings. Ofc you'll have to do that for each pointer in the slice

I don't really want to busy up this issue more as your question doesn't relate to variadic functions. If you have more questions about this please ask on the discord.

**Comment by JupiterRider** [comment]
@craig65535 @TotallyGamerJet 
My [libffi](https://github.com/JupiterRider/ffi) wrapper supports variadic functions now. See ffi.PrepCifVar.

Unfortunately ffi doesn't support MacOS yet.

**Comment by unxed** [comment]
> I should mention I'm using macOS (darwin/arm64)

See https://github.com/ebitengine/purego/issues/446

It also explains workaround I use.

**Comment by unxed** [comment]
Btw, since variadic support for Darwin ARM64 is currently hindered by the specific ABI requirement that variadic parameters reside on the stack, one approach could leverage the existing argument spilling behavior of the Go runtime. By padding a syscall with enough dummy arguments to fill all eight available ARM64 registers, any additional parameters would naturally spill onto the stack. This could potentially allow the target C function to find the variadic arguments exactly where it expects them without modifying the core assembly of purego, effectively using the register limit to our advantage.

Alternatively, a more flexible assembly trampoline than I use in #446 workaround could be designed to accept an explicit count of fixed arguments. Instead of using hardcoded stubs for every possible signature, this single generic function would dynamically calculate the necessary stack offset and alignment at runtime. This would allow a single piece of assembly code to act as a universal bridge for various variadic signatures by manually moving the appropriate number of register-based arguments onto the stack before branching to the destination function.

A more advanced but highly portable method would be the use of dynamic machine code generation. Small stubs could be constructed in memory on the fly to handle the specific register-to-stack movement required for a particular function call. By allocating an executable memory page and generating the specific ARM64 instructions needed to satisfy the Apple ABI for a given number of arguments, purego could provide universal variadic support without requiring any static assembly files or large external libraries like libffi.

**Comment by unxed** [comment]
Workaround requiring no code chage: https://github.com/unxed/pureffi
Thanks to this: https://github.com/go-webgpu/goffi/commit/8dfbc56cd4b0e91381575be82a4cdad78c289112
